lib/bakeAmbientOcclusion.js

```diff
@@ -43,20 +48,323 @@ function bakeAmbientOcclusion(gltf, options) {
+        for (var i = 0; i < vertexCount; i++) {
+            samples[i] /= counts[i];
+        }
+        parameters.allAOData = parameters.allAOData.concat(samples);
+        parameters.primitiveOrder.push(primitive);
+        parameters.primitiveNames.push(meshPrimitiveID);
+        parameters.primitiveVertexCounts.push(vertexCount);
+    }
+}
+
+function addVertexData(gltf, scene, raytracerScene) {
+    // Get all the ao vertex data together, ordered parallel with the vertex data
+    // - append all the aoBuffers from the primitives
+    // - record the order for the accessors
+    var parameters = {
+        aoBufferByPrimitive: raytracerScene.aoBufferByPrimitive,
+        primitiveOrder: [],
+        primitiveNames: [],
+        primitiveVertexCounts: [],
+        allAOData: []
+    };
+
+    NodeHelpers.forEachPrimitiveInScene(gltf, scene, concatenateAoBuffers, parameters);
+
+    var primitiveOrder = parameters.primitiveOrder;
+    var allAOData = parameters.allAOData;
+    var allAODataLength = allAOData.length;
+
+    gltf.buffers.aoBuffer = {
+        input: {
+            byteLength : allAODataLength * 4, // (FLOAT). TODO: add option to bake to shorts
+            type: 'arrayBuffer',
+            uri: 'data:'
+        },
+        extras: {
+            _pipeline: {
+                source: new Buffer(new Float32Array(allAOData).buffer)
+            }
+        }
+    };
+
+    // add buffer view
+    gltf.bufferViews.aoBufferView = {
+        buffer: 'aoBuffer',
+        byteOffset: 0,
+        byteLength: allAODataLength * 4,
+        target: WebGLConstants.ARRAY_BUFFER
+    };
+
+    // add accessor for each primitive
+    var primitiveCount = primitiveOrder.length;
+    var byteOffset = 0;
+    for (var i = 0; i < primitiveCount; i++) {
+        var primitive = primitiveOrder[i];
+        var primitiveVertexCount = parameters.primitiveVertexCounts[i];
+        var name = 'accessor_' + parameters.primitiveNames[i] + '_AO';
+        primitive.attributes.VERTEX_AO = name;
+        gltf.accessors[name] = {
+            bufferView: 'aoBufferView',
+            byteOffset: byteOffset,
+            componentType: WebGLConstants.FLOAT,
+            count: primitiveVertexCount,
+            type: 'SCALAR'
+        };
+        byteOffset += primitiveVertexCount * 4;
+    }
+}
+
+function addAoToShaders(gltf, techniquesIDsToLookAt, options) {
+    // Keep track of which shaders have been edited
+    var shaders = gltf.shaders;
+    var shadersEdited = {};
+    for (var shaderID in shaders) {
+        if (shaders.hasOwnProperty(shaderID)) {
+            shadersEdited[shaderID] = false;
+        }
+    }
+
+    // For each technique,
+    var techniquesCount = techniquesIDsToLookAt.length;
+    for (var i = 0; i < techniquesCount; i++) {
+        var technique = gltf.techniques[techniquesIDsToLookAt[i]];
+
+        var glslNewAttributes = 'attribute float a_ambientOcclusion; \n'; // snippet for adding attributes
```

Comment:
For the roadmap, we should also consider just storing this, for example, as the `w` component of the position or normal to reduce the total number of attributes, which will mean fewer WebGL calls (at least without VAOs) and being less likely to hit the limit on the maximum number of vertex attributes.

No need to do this now though.